### PR TITLE
Make Tensor::data() return type more generic.

### PR DESCRIPTION
**Pull request by xunzhang**

### CODE REVIEW

File: horovod/common/common.h [hunk 1]
```diff
@@ -86,7 +86,7 @@ class Tensor {
 public:
   virtual const MPIDataType dtype() const = 0;
   virtual const TensorShape shape() const = 0;
-  virtual const char* data() const = 0;
+  virtual const void* data() const = 0;
```
Comment (alsrgv) [inline horovod/common/common.h]: Can you align `PersistentBuffer::AccessData` to `const void*` as well?

Comment (xunzhang) [reply horovod/common/common.h]: done.

**Comment by alsrgv** [comment]
LGTM, thanks!

**Comment by alsrgv** [comment]
@xunzhang, I just compiled latest Horovod sources and we now have few warnings about pointer arithmetic:

```
    horovod/common/operations.cc: In function ‘void horovod::common::{anonymous}::PerformOperation(horovod::common::{anonymous}::TensorTable&, horovod::common::MPIResponse)’:
    horovod/common/operations.cc:824:60: warning: pointer of type ‘void *’ used in arithmetic [-Wpointer-arith]
                          cudaMemcpyAsync((void*)(buffer_data + offset),
                                                                ^
    horovod/common/operations.cc:522:25: note: in definition of macro ‘CUDA_CHECK’
         auto cuda_result = (op);                                                   \
                             ^
    horovod/common/operations.cc:853:53: warning: pointer of type ‘void *’ used in arithmetic [-Wpointer-arith]
                                          (buffer_data + offset),
                                                         ^
    horovod/common/operations.cc:522:25: note: in definition of macro ‘CUDA_CHECK’
         auto cuda_result = (op);                                                   \
                             ^
    horovod/common/operations.cc:953:48: warning: pointer of type ‘void *’ used in arithmetic [-Wpointer-arith]
                              (void*)(buffer_data + offset),
                                                    ^
    horovod/common/operations.cc:522:25: note: in definition of macro ‘CUDA_CHECK’
         auto cuda_result = (op);                                                   \
                             ^
    horovod/common/operations.cc:959:45: warning: pointer of type ‘void *’ used in arithmetic [-Wpointer-arith]
               std::memcpy((void*)(buffer_data + offset),
                                                 ^
    horovod/common/operations.cc:997:41: warning: pointer of type ‘void *’ used in arithmetic [-Wpointer-arith]
                              (buffer_data + offset),
                                             ^
    horovod/common/operations.cc:522:25: note: in definition of macro ‘CUDA_CHECK’
         auto cuda_result = (op);                                                   \
                             ^
    horovod/common/operations.cc:1003:38: warning: pointer of type ‘void *’ used in arithmetic [-Wpointer-arith]
                           (buffer_data + offset),
                                          ^
```

We should fix this.

**Pull request merged**